src/test/java/htsjdk/samtools/filter/TagFilterTest.java

```diff
@@ -61,18 +121,109 @@ public void testTagFilter(final String testName, final String tag, final List<Ob
+     */
+    @DataProvider(name = "dataIncludeNonMatchingPairedFilter")
+    private Object[][] getIncludeNonMatchingPairedTagFilterTestData()
+    {
+        return new Object[][]{
+                {"Paired Read Non Matching Basic positive test - includeReads true", ReservedTagConstants.XN, Arrays.asList(1), 1, false, true, false},
+                {"Paired Read Non Matching Multi-value positive tes - includeReads truet", ReservedTagConstants.XN, Arrays.asList(1,2,3), 1, false, true, false},
+                {"Paired Read Non Matching Incorrect value negative test - includeReads true", ReservedTagConstants.XN, Arrays.asList(1), 2, true, true, false},
+                // this is looking for value `null` which in our test cases
+                {"Paired Read Non Matching Null value negative test - includeReads true", ReservedTagConstants.XN, Arrays.asList(1), null, true, true, false},
+        };
+    }
+
+    /**
+     * Data for various sequences which may or may not match the includeReads=true default filter with paired reads.
+     */
+    @DataProvider(name = "dataIncludeMatchingPairedFilter")
+    private Object[][] getIncludeMatchingPairedTagFilterTestData()
+    {
+        return new Object[][]{
+                {"Paired Read Matching Basic positive test", ReservedTagConstants.XN, Arrays.asList(1), 1, false, true, true},
+                {"Paired Read Matching Multi-value positive test", ReservedTagConstants.XN, Arrays.asList(1,2,3), 1, false, true, true},
+                {"Paired Read Matching Incorrect value negative test", ReservedTagConstants.XN, Arrays.asList(1), 2, true, true, true},
+                // this is looking for value `null` which in our test cases
+                {"Paired Read Matching Null value negative test", ReservedTagConstants.XN, Arrays.asList(1), null, true, true, true},
+        };
+    }
+
+    // Base paired read test
+    private void testPairedReadTagFilter(final String testName, final String tag, final List<Object> validValues,
+                                         final Object testValue, final boolean expectedResult, final Boolean includeReads,
+                                         final boolean matchPairs) {
+        final TagFilter filter = new TagFilter(tag, validValues, includeReads);
+        final SAMRecordSetBuilder builder = new SAMRecordSetBuilder();
+        builder.addPair("Paired", 1, 100, 200);
+        CloseableIterator<SAMRecord> iterator = builder.iterator();
+        final SAMRecord record1 = iterator.next();
+        if (testValue != null) {
+            record1.setAttribute(tag, testValue);
+        }
+        final SAMRecord record2 = iterator.next();
+        if (matchPairs && testValue != null) {
+            record2.setAttribute(tag, testValue);
+        } else if (!matchPairs){
+            record2.setAttribute(tag, 0);
+        }
+        Assert.assertEquals(filter.filterOut(record1, record2), expectedResult, testName);

```

Comment:
shouldn't you also test `filterOut(SamRecord)` on it's own?

Reply:
Ah yes I can just refactor all of my tests and do both at once 👍